tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs

```diff
@@ -0,0 +1,613 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers;
+using MongoDB.Bson.TestHelpers.JsonDrivenTests;
+using MongoDB.Driver.Core;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Configuration;
+using MongoDB.Driver.Core.ConnectionPools;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.Helpers;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.Servers;
+using Moq;
+using Xunit;
+
+namespace MongoDB.Driver.Specifications.connection_monitoring_and_pooling
+{
+    public class ConnectionMonitoringAndPoolingTestRunner
+    {
+        #region static
+        private static readonly string[] __alwaysIgnoredEvents =
+        {
+            nameof(ConnectionPoolOpeningEvent),
+            nameof(ConnectionPoolAddingConnectionEvent),
+            nameof(ConnectionPoolAddedConnectionEvent),
+            nameof(ConnectionPoolCheckingInConnectionEvent),
+            nameof(ConnectionPoolRemovingConnectionEvent),
+            nameof(ConnectionPoolRemovedConnectionEvent),
+            nameof(ConnectionPoolClosingEvent),
+            nameof(ConnectionPoolClearingEvent),
+            nameof(ConnectionOpeningEvent),
+            nameof(ConnectionClosingEvent)
+        };
+        #endregion
+
+        [SkippableTheory]
+        [ClassData(typeof(TestCaseFactory))]
+        public void RunTestDefinition(JsonDrivenTestCase testCase)
+        {
+            var connectionMap = new ConcurrentDictionary<string, IConnection>();
+            var eventCapturer = new Lazy<EventCapturer>(isThreadSafe: true);
+            var tasks = new ConcurrentDictionary<string, Task>();
+
+            var test = testCase.Test;
+            JsonDrivenHelper.EnsureAllFieldsAreValid(test, "_path", "version", "style", "description", "poolOptions", "operations", "error", "events", "ignore", "async");
+            EnsureAvailableStyle(test);
+            ResetConnectionId();
+
+            var connectionPool = new Lazy<IConnectionPool>(
+                () =>
+                {
+                    var pool = SetupConnectionPool(test, eventCapturer.Value);
+                    pool.Initialize();
+                    return pool;
+                },
+                true);
+
+            var operations = testCase.Test.GetValue("operations").AsBsonArray;
+            var async = testCase.Test.GetValue("async").ToBoolean();
+            Exception exception = null;
+            foreach (var operation in operations.Cast<BsonDocument>())
+            {
+                ExecuteOperation(
+                    operation,
+                    eventCapturer.Value,
+                    connectionMap,
+                    tasks,
+                    connectionPool.Value,
+                    async,
+                    out exception);
```

Comment:
Would it make sense to have a single try/catch at this level so each operation doesn't have to have its own try/catch to set the exception variable?

Reply:
Why would we need to filter our own exceptions?

If an exception was thrown either it matches the expected exception or it doesn't.

Reply:
But you can use multiple try/catch if you wish. It's not stopping me from aprroving the PR.

Reply:
The idea of the `AssertError` method (at least for me) was to validate errors from the tested driver code, not from the tests itself. So, in case if we see (for example in the evergreen results) an error which contains an error like: 

    Unexpected exception has been thrown

https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/57/files#diff-a638783da40724c2c11679e9321e1829R101

It will mean that something wrong in the driver code only. And it's not related to, for example, the fact that JSON files were changed.
It's not a big deal to implement your suggestion, but I would prefer to leave it as is.

However. I noticed one minor problem when we could overwrite the exception if it has been thrown more than in one step. Fixed in the last commit.